src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
+
+                    if (tokenIndex + 5 >= tokens.Length || tokens[tokenIndex + 5] != 0)
+                    {
+                        // We've hit the next line or the end of the tokens
+                        break;
+                    }
+
+                    tokenIndex += 5;
+                }
+
+                lineToTokensDict.Add(absoluteLine, lineTokens);
+            }
+
+            // Fill in the gaps for empty lines. When looking up tokens later, this helps us differentiate
+            // between empty lines vs. lines we don't have cached tokens for
+            for (var lineIndex = range.Start.Line; lineIndex < range.End.Line; lineIndex++)
+            {
+                if (!lineToTokensDict.TryGetValue(lineIndex, out _))
+                {
+                    lineToTokensDict.Add(lineIndex, Array.Empty<int>());
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to retrieve the cached tokens for a given document/semanticVersion/range. If the cache does
+        /// not have complete tokens for the range, it's possible to return only tokens associated with a portion
+        /// of the range.
+        /// </summary>
+        /// <param name="uri">The URI associated with the cached tokens.</param>
+        /// <param name="semanticVersion">The semantic version associated with the cached tokens.</param>
+        /// <param name="requestedRange">The requested range for the desired tokens.</param>
+        /// <param name="cachedTokens">If found, contains the cached range and cached tokens.</param>
+        /// <returns></returns>
+        internal bool TryGetCachedTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range requestedRange,
+            [NotNullWhen(true)] out (Range Range, List<int> Tokens)? cachedTokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache) ||
+                !documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            var tokens = new List<int>();
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of where the cached range ends
+            var cachedRangeNumLines = 0;
+
+            // Keeps track of the current empty line count (needed for calculating line offsets)
+            var emptyLineCount = 0;
+
+            // Retrieve the cached tokens associated with the passed-in range
+            for (var line = requestedRange.Start.Line; line < requestedRange.End.Line; line++)
```

Comment:
can you raname this to currentLineNumber or something like that?  `line` makes me feel like it actually was some sort of piece of rich data.